src/GafferScene/Instancer.cpp

```diff
@@ -403,6 +393,92 @@ class Instancer::EngineData : public Data
 					throw IECore::Exception( fmt::format( "Context primitive variable for \"{}\" is not a correctly sized Vertex primitive variable", v.name.string() ) );
 				}
 			}
+
+			if( !m_numValidPrototypes )
+			{
+				// We don't need to build m_instancesForPrototype if we're not outputting any prototypes
+				return;
+			}
+
+			int constantPrototypeIndex = -1;
+			if( !m_prototypeIndices )
+			{
+				constantPrototypeIndex = m_prototypeIndexRemap[ 0 ];
+				if( constantPrototypeIndex == -1 )
+				{
+					// If we have no indices to specify other prototypes, and the first prototype is
+					// invalid, we're not going to output anything, and can early exit
+					return;
+				}
+			}
+
+			std::vector< std::vector<int> > instancesForPrototypeIndex( m_numPrototypes );
+
+			if( constantPrototypeIndex == -1 || m_ids )
+			{
+				if( m_ids )
+				{
+					if( constantPrototypeIndex != -1 )
+					{
+						instancesForPrototypeIndex[ constantPrototypeIndex ].reserve( numPoints() );
+					}
+
+					for( size_t i = 0, e = numPoints(); i < e; ++i )
+					{
+						int id = (*m_ids)[i];
+						if( m_idsToPointIndices.try_emplace( id, i ).second )
```

Comment:
As part of the aforementioned refactor, I've decoupled this from computing pointIndicesForPrototoypeIndex, which I think will make you happier with this code - the part I hadn't realized when I first wrote it is that I can just store a flag for whether there are duplicates, and if they're aren't, I don't need to repeat the expensive m_idsToPointIndices map lookup while computing pointIndicesForPrototoypeIndex - it was this map lookup that has a measurable impact to do twice.

In term of where the initialization happens, I think it would actually make sense for it to happen lower down in this function, after initAttributes, and after the early out for "no valid prototypes", but I've moved it back to where it was for now for convenience in reviewing.